src/App.tsx

```diff
@@ -1,18 +1,206 @@
+
+  const [addingError, setAddingError] = useState(false);
+  const [deletingError, setDeletingError] = useState(false);
+  const [addingNewTodo, setAddingNewTodo] = useState(false);
+  const [updatingError, setUpdatingError] = useState(false);
+
+  const [toggleAll, setToggleAll] = useState(false);
+
+  const increaseCounter = () => {
+    setCounter(prevState => prevState + 1);
+  };
+
+  const filterTodos = (filterBy: FilterBy) => {
+    if (todos) {
+      setVisibleTodos(todos.filter((todo: Todo) => {
+        switch (filterBy) {
+          case FilterBy.ACTIVE:
+            return !todo.completed;
+          case FilterBy.COMPLETED:
+            return todo.completed;
+          default:
+            return true;
+        }
+      }));
+    }
+  };
+
+  const todosLength = () => {
+    return todos?.filter(todo => !todo.completed).length || '0';
+  };
+
+  const todosCompletedLength = () => {
+    return todos?.filter(todo => todo.completed).length;
+  };
+
+  const getTodosFromApi = useCallback(async () => {
+    try {
+      if (user) {
+        setIsAdding(true);
+        const apiTodos = await getTodos(user.id).then(gotTodos => gotTodos);
+
+        if (apiTodos.length === 0) {
+          setTodos([]);
+          setVisibleTodos([]);
+          setIsAdding(false);
+
+          return;
+        }
+
+        setTodos(apiTodos);
+        setVisibleTodos(apiTodos);
+        setIsAdding(false);
+      }
+    } catch {
+      setAddingError(true);
+    }
+  }, []);
+
+  const handleFilter = (filterBy: FilterBy) => {
+    filterTodos(filterBy);
+    setSelectedFilter(filterBy);
+  };
+
   useEffect(() => {
     // focus the element with `ref={newTodoField}`
     if (newTodoField.current) {
       newTodoField.current.focus();
     }
-  }, []);
+
+    getTodosFromApi();
+  }, [counter]);
+
+  const debounceCallback = useCallback(
+    () => {
+      setEmptyTitleError(false);
+      setAddingError(false);
+      setUpdatingError(false);
+    }, [],
+  );
+
+  const onSetError = useMemo(
+    () => debounce(debounceCallback, 3000),
+    [],
+  );
+
+  const closeError = () => {
+    setEmptyTitleError(false);
+    setAddingError(false);
+    setDeletingError(false);
+    setUpdatingError(false);
+  };
+
+  const handleOnChangeNewTodoTitle = (event: ChangeEvent<HTMLInputElement>) => {
+    setNewTodoTitle(event.target.value);
+  };
+
+  const handleOnSubmitNewTodo = async (event: FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    setAddingNewTodo(true);
+    if (newTodoTitle.trim() === '') {
+      setEmptyTitleError(true);
+
+      return;
+    }
+
+    if (user) {
+      await addNewTodo({
+        userId: user.id,
+        title: newTodoTitle,
+        completed: false,
+      });
+
+      setCounter((prevState) => prevState + 1);
+    }
+
+    setNewTodoTitle('');
+    setAddingNewTodo(false);
+  };
+
+  const onDeleteTodo = async (id: number) => {
+    await deleteTodo(id);
+    increaseCounter();
```

Comment:
for what purpose did you add counter value? can we calculate it according to the total length of todos?